### PR TITLE
ARC fixes. closes #50

### LINKED ISSUE #50: OF - Transition to ARC
Opening this issue here to notify a transition to ARC in the next OF Release
Some more info here
https://github.com/openframeworks/openFrameworks/issues/6848
https://github.com/openframeworks/openFrameworks/issues/6848#issuecomment-1064459744
Thanks

### PR DESCRIPTION
**Pull request by ofTheo**
This is just the changes needed for ofxSyphon to work with ARC ( automatic reference counting ), which will be enabled on the next OF release ( and is already enabled in the nightly builds ). 

Confirmed working with the nightly downloads. 

Would love to get @2bbb and @danomatika 's take on this. 

Especially the way I am doing cleanup of the void pointers which we use for the client and server. 

https://github.com/astellato/ofxSyphon/compare/master...ofTheo:syphon-arc?expand=1#diff-ab7cfefbfb00d042e397c6984ba2f3f2d75db565a6a1e4e832a942891bd4f4d3R58-R63

NOTE: don't merge this to master until the next OF release.

### CODE REVIEW

**Comment by dimitre** [comment]
I've just tested this one and everything seems to be working great!

File: src/ofxSyphonClient.mm [hunk 1]
```diff
@@ -46,32 +41,36 @@
...
 }
 
+
 void ofxSyphonClient::setup()
 {
-    // Need pool
-    NSAutoreleasePool* pool = [[NSAutoreleasePool alloc] init];
-        
-	mClient = [[SyphonNameboundClient alloc] initWithContext:CGLGetCurrentContext()];
+    //sketchy as we transfer ownership to void * and that means we have to manage the memory ourselves so in destructor we are going to make it back into an obj-c object and transfer back so ARC can do its thing.
```
Comment (danomatika) [inline src/ofxSyphonClient.mm:69]: Different indentation.

File: src/ofxSyphonServer.h [hunk 2]
```diff
@@ -20,4 +20,5 @@ class ofxSyphonServer {
     void publishTexture(GLuint id, GLenum target, GLsizei width, GLsizei height, bool isFlipped);
 	protected:
 	void *mSyphon;
+    void cleanup();
```
Comment (danomatika) [inline src/ofxSyphonServer.h:23]: Different indentation.

File: src/ofxSyphonServer.mm [hunk 3]
```diff
@@ -17,44 +17,43 @@
...
 {
-    NSAutoreleasePool* pool = [[NSAutoreleasePool alloc] init];
-    
 	NSString *title = [NSString stringWithCString:n.c_str()
 										 encoding:[NSString defaultCStringEncoding]];
 	
 	if (!mSyphon)
 	{
-		mSyphon = [[SyphonOpenGLServer alloc] initWithName:title context:CGLGetCurrentContext() options:nil];
+        //sketchy as we transfer ownership to void * and that means we have to manage the memory ourselves so in destructor we are going to make it back into an obj-c object and transfer back so ARC can do its thing.
```
Comment (danomatika) [inline src/ofxSyphonServer.mm:41]: Different indentation.

**Comment by danomatika** [comment]
Looks fine to me, other than a few indentation issues.

**Comment by ofTheo** [comment]
Thanks @danomatika - I can fix the indentation stuff, only noticed it in the PR diff.

**Comment by ofTheo** [comment]
@bangnoise - Thanks! 
I guess the void * is needed to avoid issues with including Obj-C includes from pure C++.

If there is a way around that, then the copy constructors, destructors etc wouldn't even be needed as with ARC all the objects would just maintain their own reference count. 

I don't know too much about auto release pools but looking at:
https://stackoverflow.com/questions/9086913/why-is-autoreleasepool-still-needed-with-arc 

It looks like it is useful when you want stuff to be released at the end of the function as opposed to at some point when the main autorelease pool decides. So maybe I could add back in @autorelease blocks to where the pools where before?

Anyway, def happy to have some extra eyes on this. 
@2bbb helped us with [our ARC transition](https://github.com/openframeworks/openFrameworks/pull/6889/files).

**Comment by danomatika** [comment]
> It looks like it is useful when you want stuff to be released at the end of the function as opposed to at some point when the main autorelease pool decides. So maybe I could add back in @autorelease blocks to where the pools where before?

In full Obj-C projects, the main function has an autorelease pool basically wrapping everything, so ARC transitions there don't need any additional pools, as far as I remember. Here, though, we have our own C++ main without a pool, so I think you are right in that there needs to be one within these classes. I'm not an expert though, having learned Obj-C *thankfully* just as ARC was released. :)

**Review by bangnoise — CHANGES_REQUESTED** [review]
see in-line comments

**Comment by bangnoise** [comment]
Apologies, I didn't press submit on my review, but GitHub confusingly shows my comments here (for me) before submitting. I meant the in-code comments to precede my last comment above - hopefully it makes a bit more sense now.

File: src/ofxSyphonClient.mm [hunk 4]
```diff
@@ -19,17 +19,12 @@
...
-    NSAutoreleasePool* pool = [[NSAutoreleasePool alloc] init];
-
-    [(SyphonNameboundClient*)mClient release];
-    mClient = nil;
-    
-    [pool drain];
+    cleanup();
 }
 
 ofxSyphonClient::ofxSyphonClient(ofxSyphonClient const& s) :
```
[the remark this answers is not in the document]
Comment (bangnoise) [reply src/ofxSyphonClient.mm]: I've done this in 4b82b56f7e20ed78832f1dbbd5a2fc579547dc53

File: src/ofxSyphonClient.mm [hunk 5]
```diff
@@ -46,32 +41,36 @@
     {
         return *this;
     }
-    NSAutoreleasePool* pool = [[NSAutoreleasePool alloc] init];
```
[the remark this answers is not in the document]
Comment (bangnoise) [reply src/ofxSyphonClient.mm:49]: I've restored the autorelease pools in 4b82b56f7e20ed78832f1dbbd5a2fc579547dc53

**Review by bangnoise — APPROVED** [review]

**Comment by bangnoise** [comment]
I've added a basic wrapper C++ class which takes care of the NSObject lifetime - this fixes a lot of issues where ARC would lose track of objects we were storing in `void*` member variables, and makes `ofxSyphonServer` and `ofxSyphonClient` simpler - they no longer need custom copy and assignment operators.

I've restored the deleted autorelease pools so we don't leak Objective C objects when called from C++ code without an autorelease pool in place.

I've also updated the Xcode projects using a Project Generator from OF nightly.

**Comment by danomatika** [comment]
@bangnoise Thanks for taking the time to integrate this. I see now that I also need to integrate pools within my PGMidi Obj-C class wrappers for ofxMidi...

**Comment by ofTheo** [comment]
Hey @bangnoise, this approach looks great! 
Great to get rid of the void *.  Def makes the whole ARC approach less hack-y. 

I gave it a quick test and it all seems to work well. 
@dimitre maybe try pulling in these changes and see if you notice any issues with your setup?

**Comment by dimitre** [comment]
I've just pulled fc94e9a and tested here in my setup, using syphon recorder and everything is working great here.
using 4 syphon outputs in this setup. 
Screen Shot 2022-09-16 at 13 08 18

**Comment by danomatika** [comment]
PIMPL is a pretty well used idiom. I’d think ARC wouldn’t have a problem with it…

enohp ym morf tnes
-----------
Dan Wilcox
danomatika.com
robotcowboy.com

**Pull request merged**